source/base/vectorization.cc

```diff
@@ -18,7 +18,9 @@
 DEAL_II_NAMESPACE_OPEN
 
 #if DEAL_II_COMPILER_VECTORIZATION_LEVEL >= 1
+template <>
 const unsigned int VectorizedArray<double>::n_array_elements;
+template <>
 const unsigned int VectorizedArray<float>::n_array_elements;
```

Comment:
Hm, thinking about this again, I'm not sure this is correct. The question is whether `VectorizedArray` is a template or not. If, for example, you have `DEAL_II_COMPILER_VECTORIZATION_LEVEL >= 3`, then there is
```

#if DEAL_II_COMPILER_VECTORIZATION_LEVEL >= 3 && defined(__AVX512F__)

/**
 * Specialization of VectorizedArray class for double and AVX-512.
 */
template <>
class VectorizedArray<double>
{
...
```
and so the class is not a template, and the definition of the member variable should not be preceded by `template <>` because it's not a specialization. The same applies if `DEAL_II_COMPILER_VECTORIZATION_LEVEL >= 2 && defined(__AVX__)` and `DEAL_II_COMPILER_VECTORIZATION_LEVEL >= 1 && defined(__SSE2__)` and now `DEAL_II_VECTORIZATION_LEVEL >= 1 && defined(__ALTIVEC__) && defined(__VSX__)`. So I guess this also covers your case.

Why do you need the `template <>` here?

Reply:
I see now, it is incorrect as the n_array_elements is already defined in the template specializations in the vectorization.h file for `DEAL_II_COMPILER_VECTORIZATION_LEVEL >=1` 
How about this then?
```
#if DEAL_II_COMPILER_VECTORIZATION_LEVEL < 1
template <> const unsigned int VectorizedArray<double>::n_array_elements;
template <> const unsigned int VectorizedArray<float>::n_array_elements;
#endif
```

Reply:
I agree with @bangerth here. The last commit gives me
```
../include/deal.II/base/vectorization.h:2208:29: warning: too many template headers for ‘dealii::VectorizedArray<double>::n_array_elements’ (should be 0)
   static const unsigned int n_array_elements = 2; 
```
for SSE2.

If you get the compiler error you show, that means that the template specialization is not defined which implies that either `__ALTIVEC__` or `__VSX__` is not defined. Can you confirm this?

Reply:
@dsambit -- yes, I think that is correct.

Reply:
> This compiled fine in the altivec case.

That is confusing. The error you show above suggests `DEAL_II_COMPILER_VECTORIZATION_LEVEL >= 1` but your last suggestions would only change anything for  `DEAL_II_COMPILER_VECTORIZATION_LEVEL == 0`.
What is `DEAL_II_COMPILER_VECTORIZATION_LEVEL` for you?

Reply:
@masterleinad `DEAL_II_COMPILER_VECTORIZATION_LEVEL` for me is 1.

Reply:
I still believe that you don't have the correct specializations for `double` and `float`. What is `n_array_elements`?

Assuming my assumption is true, I understand that what you are doing works for you because you just disable the definition of `n_array_elements` for the (non-existing) specializations of `VectorizedArray`. 
This might (no diagnostic required) lead to errors like shown in  https://groups.google.com/forum/#!topic/dealii/fJwVDhHORnY if `n_array_elements` is odr-used.

Reply:
You are right- this is still incorrect. Previously I was not compiling the examples. If I compile the examples- I get the same errors as referenced in https://groups.google.com/forum/#!topic/dealii/fJwVDhHORnY
```
[ 89%] Linking CXX executable ../bin/step-56.debug
../lib/libdeal_II.g.so.9.1.0-pre: undefined reference to `dealii::VectorizedArray<double>::n_array_elements'
../lib/libdeal_II.g.so.9.1.0-pre: undefined reference to `dealii::VectorizedArray<float>::n_array_elements'
```